enhancements/cluster-logging/elasticsearch-hot-warm-architecture.md

```diff
@@ -0,0 +1,156 @@
+
+## Proposal
+
+Currently we offer a mechanism for users to configure their index lifecycles based on two states: hot and delete. By default indices are created and stay in the hot state, where they can be rolled over based on a configuration, however they stay in a "hot" state until they are eventually deleted based on the delete policy.
+
+Elasticsearch offers mechanisms to move indices between hot, warm, cold, and delete and discusses approaches for this [in a blog article](https://www.elastic.co/blog/implementing-hot-warm-cold-in-elasticsearch-with-index-lifecycle-management). However, this functionality is a x-pack feature and we do not support these. [Opendistro](https://opendistro.github.io/for-elasticsearch-docs/docs/ism/) offers a way to do this as well, however their plugins require ES7. We achieve our current functionality using cronjobs -- one for delete and one for rollover.
+
+Within the context of Elasticsearch ILM:
+1) Moving the index into a Warm state means the index is force merged to use less primary shards, its replica count is reduced, and the index priority is also reduced. The idea of this being that these indices are not being written to, may not be queried against as much so they do not need to be as distributed, and can be a lower priority when it comes to index recovery (meaning they would be recovered after high priority indices are).
+
+1) Moving into a Cold state further reduces its index priority to be lower than Warm (but still higher than delete) and, if one was using specific nodes to house indices of certain state it would move from Warm to Cold nodes. This is different than Closing an index because the index can still be queried (if one were to Close an index it would need to be Opened first before it can be searched).
+
+### User Stories
+
+#### As an EO admin, I want to be able to configure a warm policy for my indices so that I can keep searchable data around longer and take up less resources
+
+Currently we keep indices at their original primary and replica shard count which comes at a cost in ES. Reducing these can help us increase our retention policy.
+
+### Implementation Details
+
+#### Assumptions
+
+* This will be going out with our ES6 operator which means it will be a part of our ES 6.x cluster
+
+* If not specified to be separate Warm and Rollover will be synonomous
+
+#### Security
+
+No additional security concerns need to be addressed by this.
+
+### Risks and Mitigations
+
+## Design Details
+
+Currently the API reads differently for Hot and Delete. Hot uses the MaxAge of an index and Delete the MinAge. For the sake of this proposal Warm will follow the same convention as Delete, a later iteration can seek to align all phases to use the same terminology (non-goal).
+
+Given that having three cronjobs for a single index (with all three policies defined) may be noisy, as part of this we should also consolidate the index management cronjobs into a single one.
```

Comment:
clarify 'noisy'.  Consideration needs to be taken not to block an operation when another fails.  Singles were implemented to define with the idea to do one thing only and do it well (even if we still are striving to acheive this)

Reply:
I'm sure we can re-write our cronjob for one operation to not block another if it fails...

Reply:
> I'm sure we can re-write our cronjob for one operation to not block another if it fails... to me if we have 9 cron jobs out there, that is too much..

It does no sound like there is any concrete criteria to base this assessment.

>  that is a lot of pods we leave behind, that is a lot of cronjobs to maintain. 

One can always remove the "failed" pods

> it seems unnecessary. we could even consider consolidating it all into a single cronjob since there is a single polling schedule for all the operations.

Ultimately this can be resolved with various designs but the most clear cut "this is what I do" with limited execution paths is more maintainable IMO

Reply:
I think we're going to have to agree to disagree. I feel 9 cronjobs is a poor UX and creates excess noise. I feel we have unnecessary redundancy that doesn't actually buy us anything that we can't solve with defensive coding.

I want to shift the paradigm from "I do this one very specific thing for this pattern" to "I do management for this pattern"